HCL2 Parse required plugin block

azr opened the pull request:
This simply does the parsing and some testing.

This is merges into #10277

park-cheif5 reviewed (commented):
What do you think  🤔 💭 😜

nywilken commented on packer/plugin_ext_unix.go:
```diff
@@ -0,0 +1,9 @@
+import pluginsdk "github.com/hashicorp/packer-plugin-sdk/plugin"
+
+var (
+	PluginFileExtension = "_x" + pluginsdk.APIVersion // OS-Specific plugin file extention
```
I might have missed this. But is the API version required in the filename?

azr replied:
Yes

azr replied:
( Terraform does the same )

nywilken reviewed (commented):
This looks really good. I'm a bit fan of the new types and how the code is broken out. The github pkg also looks really interesting. 

I left a couple of questions.

SwampDragons commented on command/core_wrapper.go:
```diff
@@ -23,3 +24,13 @@ func (c *CoreWrapper) Initialize() hcl.Diagnostics {
+func (c *CoreWrapper) PluginRequirements() (plugingetter.Requirements, hcl.Diagnostics) {
+	return nil, hcl.Diagnostics{
+		&hcl.Diagnostic{
+			Summary:  "Packer init is not supported in Packer - JSON",
```
I like this suggestion -- we should also provide a docs link for upgrading to HCL

SwampDragons replied:
This is a good reminder that our "extending" page needs to have details for how to install for json AND for HCL since the paths are now different

azr replied:
Yup, that's way better ! Thanks.

I'll definitely add this to docs !

azr replied:
This is done now.

SwampDragons commented on command/init.go:
```diff
@@ -0,0 +1,140 @@
+			continue
+		}
+
+		newInstall, err := pluginRequirement.InstallLatest(plugingetter.InstallOptions{
```
looks like currently this doesn't take compatability restrictions into account. We skipping that for now?

azr replied:
`opts.BinaryInstallationOptions` contains the os/version/api-version settings in there.
https://github.com/hashicorp/packer/blob/878fc823a6e8ec6b45cbfab841bbfba98405a4f0/command/init.go#L97-L101

Now the installer will get the latest possible version matching the `required_plugin` and fail if the binary that matches all of these settings is not found.

SwampDragons commented on config_test.go:
```diff
@@ -48,27 +48,30 @@ func TestLoadExternalComponentsFromConfig(t *testing.T) {
 
-	if len(cfg.Provisioners) != 1 || !cfg.Provisioners.Has("super-shell") {
-		t.Errorf("failed to load external provisioners; got %v as the resulting config", cfg.Provisioners)
+	if len(cfg.Plugins.Provisioners) != 1 || !cfg.Plugins.Provisioners.Has("super-shell") {
```
I like that you've moved these into a plugins substruct -- makes it easier to mentally distinguish the template config from the plugins themselves.

SwampDragons commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+	if p.Namespace != "" {
+		ns = p.Namespace
+	}
+	return "/" + ns + "/packer-plugin-" + p.Type
```
it feels a little weird to me to have a relative path prefixed with /

azr replied:
Oh, good catch, I actually remember wanting to remove that !

```suggestion
	return ns + "/packer-plugin-" + p.Type
```

SwampDragons commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+// the plugin is using the default hostname, the hostname is omitted.
+func (p *Plugin) ForDisplay() string {
+	parts := []string{}
+	if p.Hostname != DefaultPluginHost {
```
what if p.Hostname is empty?

SwampDragons replied:
Nevermind, this looks like it is never empty because of validation below.

azr replied:
Yes that should be the case !

SwampDragons commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+// rather than some similar rules defined locally, because the hostname part
+// of an addrs.Plugin is already a hostname and it's ideal to use exactly
+// the same case folding and normalization rules for all of the parts.
+//
```
thanks for this docstring. super useful.

azr replied:
Kudos to the Terraform team here ! 🙂

SwampDragons commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+	}
+
+	// We don't allow names containing multiple consecutive dashes, just as
+	// a matter of preference: they look weird, confusing, or incorrect.
```
nit: can we skip the use of "weird"? its connotation can be judgmental.

azr replied:
```suggestion
	// a matter of preference: they look confusing, or incorrect.
```

SwampDragons commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+	// packer-plugin- prefix to help them self-correct.
+	const redundantPrefix = "packer-"
+	const userErrorPrefix = "packer-plugin-"
+	if strings.HasPrefix(ret.Type, redundantPrefix) {
```
I'm also struggling here. Let's use a repo that used to be pretty common: 

https://github.com/jetbrains-infra/packer-builder-vsphere

The repo is named that because it's also what it should be named on install.

I _think_ this code is to tell people they needn't name their repos things like packer-plugin-party but I don't see how this is bad, and even if it is why this is such a bad thing we need to enforce against it.

azr replied:
This pkg is mainly taken out from Terraform and this error case made sense to me so I left it.

It's in the case someone wants to require a packer plugin and names it like so:

```hcl
packer {
    required_plugins {
        packer-plugin-amazon = ">= v0" // err: should be "amazon" instead of packer-plugin-amazon
        amazon-v1 = {
            source  = "packer-plugin-amazon" // should be "amazon"
            version = ">= v1"
        }
       // ...
   }
}
```

I think this is really nice for consistency across packer configs and with Terraform.

SwampDragons replied:
aha, so the name being called in the _template_ should not contain "packer-plugin" but we're not nitpicking the name of the repo or whatever. 

Let's make sure to link this error to a docs page with an explanation like you gave above, because that makes it much clearer.

SwampDragons commented on hcl2template/parser.go:
```diff
@@ -140,16 +136,34 @@ func (p *Parser) Parse(filename string, varFiles []string, argVars map[string]st
 	}
 
+	// Decode required_plugins blocks
+	// and create implicit required_plugins blocks
```
I think this comment could be expanded. What does it mean to be an implicit required_plugin block? One Packer generates to find HC-maintained plugins for backwards compat?

azr replied:
So currently my thinking is that an implicit `required_plugins` happens when you do:

```hcl
source "amazon-ebs" "example" { ... }
```

And no `amazon-ebs` builder can be found. This will then be the equivalent of having

```hcl
packer {
    required_plugins {
        amazon = "latest"
```

( Also note that using `latest` in a normal scenario does not work and packer will ask you to set a version )

SwampDragons commented on hcl2template/plugin.go:
```diff
@@ -0,0 +1,101 @@
+			if previouslySeenBlock, found := uniq[name]; found {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  fmt.Sprintf("Duplicate required_plugin.%q block", name),
```
👍 good explanation

SwampDragons replied:
we're gonna want tests for all this logic

azr replied:
Thanks ! Yes Totally ! Leaving this unresolved as a reminder 🙂

SwampDragons commented on hcl2template/plugin.go:
```diff
@@ -0,0 +1,101 @@
+			continue
+		}
+		log.Printf("[TRACE] Found the following %q installations: %v", pluginRequirement.Identifier.ForDisplay(), installs)
+		install := installs[len(installs)-1]
```
this is grabbing the latest installation because installs is sorted by version?

azr replied:
Yes !

azr replied:
Note that this _could_ have an unexpected behaviour on versions with same semver but with a named tag, ex:

```
v1.2.3
v1.2.3-BETA
v1.2.3-BETA-2
```

I haven't tested on that case too much.

azr replied:
I'll change its name to `sortedInstalls`, so it's easier to read.

SwampDragons commented on hcl2template/testdata/complete/build.pkr.hcl:
```diff
@@ -1,4 +1,33 @@
 
+packer {
```
nice. Can we add an invalid file that tests to make sure if there are duplicates we error?

azr replied:
Done !

SwampDragons commented on hcl2template/types.required_plugins.go:
```diff
@@ -0,0 +1,224 @@
+func (cfg *PackerConfig) decodeImplicitRequiredPluginsBlocks(f *hcl.File) hcl.Diagnostics {
+	// when a plugin is used but not defined in the required plugin blocks, it
+	// is 'implicitly used'. Here we read common configuration blocks to try to
+	// guess plugins.
```
so is this how the json configs will work?

azr replied:
Yes, I think so !

azr replied:
Though we'd have to implement that part JSON side for it to work.

azr replied:
My thinking here would be to release packer with `json + plugin-getter + plugins`, then release packer with `plugin-getter` ( so deprecate json and remove core plugins )

SwampDragons commented on hcl2template/types.required_plugins.go:
```diff
@@ -0,0 +1,224 @@
+						diags = append(diags, &hcl.Diagnostic{
+							Severity: hcl.DiagError,
+							Summary:  "Invalid version constraint",
+							Detail:   "This string does not use correct version constraint syntax.",
```
can we link to docs on correct syntax?

azr replied:
```suggestion
							Detail:   "This string does not use correct version constraint syntax. See https://www.packer.io/docs/from-1.5/blocks/packer#version-constraint-syntax for docs.",
```

SwampDragons commented on hcl2template/types.required_plugins.go:
```diff
@@ -0,0 +1,224 @@
+					diags = append(diags, &hcl.Diagnostic{
+						Severity: hcl.DiagError,
+						Summary:  "Invalid source",
+						Detail:   "Source must be specified as a string.",
```
let's give an example source string here

azr replied:
```suggestion
						Detail:   "Source must be specified as a string. For example: " + `source = "amazon"`,
```

SwampDragons replied:
👍

SwampDragons commented on main.go:
```diff
@@ -301,20 +297,15 @@ func extractMachineReadable(args []string) ([]string, bool) {
+	config.Plugins = &packer.PluginConfig{
+		PluginMinPort:      10000,
+		PluginMaxPort:      25000,
+		KnownPluginFolders: packer.PluginFolders("."),
```
this is cool. We should probably also use knownpluginfolders for json configs too

azr replied:
Yes !

SwampDragons commented on packer/plugin-getter/github/getter.go:
```diff
@@ -0,0 +1,181 @@
+package github
```
TODO: add gpg verification?

azr replied:
So, the way I understood it from Paul is that the GPG verification happens when we crawl a plugin repo not on installation. Basically maintainers send us ( Hashicorp ) a GPG key, and when a new version is released: We check if the GPG signature of the checksum file is good, if so we allow the download of the official plugin.
But during installation only the checksum file should be verified.

Is this correct @sylviamoss ?

sylviamoss replied:
The GPG is verified before pushing the plugin release to https://releases.hashicorp.com. Whatever is on our releases page, is verified and re-signed by HashiCorp. 

For Packer init, when downloading directly from GitHub, such verification won't happen.

azr replied:
Thanks Moss 🙂

SwampDragons commented on packer/plugin-getter/plugins.go:
```diff
@@ -0,0 +1,421 @@
+// of required plugins is generated from a config file. From it we check what
+// is actually installed and what needs to happen to get in the desired state.
+type Requirement struct {
+	// How the user named the plugin
```
```suggestion
	// name to use when using the plugin in the Packer HCL config
```

azr replied:
I made it a bit longer:
https://github.com/hashicorp/packer/blob/d8a93c61fad6d0b2d2539b1325c2803fea2de30d/packer/plugin-getter/plugins.go#L25-L29

I hope it's good !

SwampDragons commented on packer/plugin-getter/plugins.go:
```diff
@@ -0,0 +1,421 @@
+	// How the user named the plugin
+	Accessor string
+
+	// Something like github.com/hashicorp/packer-plugin-amazon
```
we're probably gonna have people trying to reach plugins on their local filesystem

azr replied:
Yup ! This should still be okay, packer will ask for a plugin named `github.com/hashicorp/packer-plugin-amazon` and a local file system plugin-getter should be able to know what to do with that. 🙂

SwampDragons commented on packer/plugin-getter/plugins.go:
```diff
@@ -0,0 +1,421 @@
+	return res, nil
+}
+
+// InstallList is a list of installs
```
this could use clarification: list of things already installed, or list of things that we need to install?

azr replied:
Yup !

SwampDragons commented on packer/plugin-getter/plugins.go:
```diff
@@ -0,0 +1,421 @@
+	Getters []Getter
+
+	// Any downloaded binary and checksum file will be put in this folder.
+	InFolders []string
```
if there are more than one folder defined, how do we choose what folder to put the downloaded binary into?

SwampDragons reviewed (commented):
Wow, this looks fantastic. I got pretty far in and ran out of time to finish the review, so I wanted to comment with my current review notes instead of waiting until tomorrow. I'll pick up from packer/plugin-getter/plugins.go in the morning.

azr commented on command/cli.go:
```diff
@@ -93,6 +93,18 @@ type BuildArgs struct {
 }
 
+func (ia *InitArgs) AddFlagSets(flags *flag.FlagSet) {
+	flags.BoolVar(&ia.Upgrade, "upgrade", false, "Wether to upgrade any already present binary.")
```
I really like your second suggestion :

```suggestion
	flags.BoolVar(&ia.Upgrade, "upgrade", false, "upgrade any present plugin to the highest allowed version.")
```

azr commented on command/init.go:
```diff
@@ -0,0 +1,140 @@
+		log.Printf("[TRACE] for plugin %s found %d matching installation(s)", pluginRequirement.Identifier.ForDisplay(), len(installs))
+
+		if len(installs) > 0 && cla.Upgrade == false {
+			continue
```
Yes ! The latest that matches the required_plugin version settings.

azr commented:
Hello team ! Many thanks for reviewing :D I'll be resolving the things that are done or ended and leave the TODO things open.

azr commented on hcl2template/parser.go:
```diff
@@ -140,16 +136,34 @@ func (p *Parser) Parse(filename string, varFiles []string, argVars map[string]st
 		return cfg, diags
 	}
 
+	// Decode required_plugins blocks
+	// and create implicit required_plugins blocks
```
```suggestion
	// Decode required_plugins blocks and create implicit required_plugins
	// blocks. Implicit required_plugins blocks happen when a builder or another
	// plugin cannot be found, for example in the if one uses :
	//  source "amazon-ebs" "example" { ... }
	// And no `amazon-ebs` builder can be found. This will then be the
	// equivalent of having :
	//  packer {
	//    required_plugins {
	//      amazon = "latest"
	//    }
	//  }
	// Note: using `latest` ( or actually an empty string ) in a config file
	// does not work and packer will ask you to pick a version
```

azr commented on packer/run_interfaces.go:
```diff
@@ -30,6 +31,7 @@ type Evaluator interface {
 // The packer.Handler handles all Packer things.
 type Handler interface {
 	Initialize() hcl.Diagnostics
+	PluginRequirements() (plugingetter.Requirements, hcl.Diagnostics)
```
TODO: document this interface more

SwampDragons commented on packer/plugin.go:
```diff
@@ -65,7 +66,7 @@ func (c *Config) Discover() error {
 	}
 
-	// Next, look in the default plugins directory inside the configdir/.packer.d/plugins.
+	// Next, look in the default plugins directory inside the configdir/.d/plugins.
```
why are we changing this location?

azr replied:
woops, that's a mistake !
```suggestion
	// Next, look in the default plugins directory inside the configdir/.packer.d/plugins.
```

SwampDragons reviewed (commented):
Looks like I pretty much got through all the meat of the PR yesterday -- this is looking great! Let me know when you're ready for me to pull and play with it.

azr commented:
Yay it builds ! Now to merge the main branch back here and fix some conflicts :D

azr commented on hcl2template/parser.go:
```diff
@@ -269,21 +298,17 @@ func (cfg *PackerConfig) Initialize() hcl.Diagnostics {
+func (p *Parser) parseConfig(f *hcl.File, cfg *PackerConfig) hcl.Diagnostics {
 	var diags hcl.Diagnostics
 
-	body := dynblock.Expand(f.Body, cfg.EvalContext(nil))
```
This no happens in the initialise phase

sylviamoss commented on command/init.go:
```diff
@@ -0,0 +1,159 @@
+		}
+
+		newInstall, err := pluginRequirement.InstallLatest(plugingetter.InstallOptions{
+			InFolders:                 c.Meta.CoreConfig.Components.PluginConfig.KnownPluginFolders,
```
```suggestion
			InFolders:                 opts.FromFolders,
```
To keep consistent with the bellow line.

azr replied:
Yup, makes sense ! Thanks !

sylviamoss commented on packer/plugin-getter/plugins.go:
```diff
@@ -0,0 +1,432 @@
+				continue
+			}
+
+			// base name could look like packer-plugin-amazon_v1.2.3_darwin_amd64_x4
```
when the version is added to the string here?

azr commented:
Head's up, I added lots of docs here https://packer-j4p3zh2nv.vercel.app/docs/plugins, I moved the 'Environment Variables' and 'Core Configuration' page into a 'Configuring Packer' page that contains docs about plugins too. I did so because I found it hard to understand what did what exactly. Please provide feedback 🙂 .

Also the link checker seems to giving me errors for links that exist in this new PR but not on packer.io, but is it possible to make it give me errors for links I broke ? I feels like it's upside down. I'll try to see.

SwampDragons commented:
I like the website refactor. It looks a lot better.

SwampDragons commented:
Love the home directory as a table. 
I pushed a tweak to the link checker so it won't scream at us. We can remove the exclusions once we've deployed after the v1.7 release.

sylviamoss commented on hcl2template/parser.go:
```diff
@@ -147,16 +140,38 @@ func (p *Parser) Parse(filename string, varFiles []string, argVars map[string]st
+			diags = append(diags, cfg.decodeRequiredPluginsBlock(file)...)
+		}
+		for _, file := range files {
+			diags = append(diags, cfg.decodeImplicitRequiredPluginsBlocks(file)...)
```
This only works for the plugins we maintain, right?

azr replied:
Well, I think that this 'guesser' should have as little info as possible and cover all already 'incore' plugins.

azr replied:
( nothing works there, this is mainly TODO, I just wanted to put the wiring there for later. )

sylviamoss commented on hcl2template/plugin.go:
```diff
@@ -0,0 +1,178 @@
+	return diags
+}
+
+func (cfg *PackerConfig) initializeBlocks() hcl.Diagnostics {
```
Would `initializeBuildBlocks` be a better name?

azr replied:
I really had a hard time naming that function and am still not very happy with the name but `Build` is kind of a meaningful word here. 
Maybe `initializeHCL2Blocks` ?  Maybe `verifyThatAllUsedPluginsDoExistAndExpandDynamicBodies` ? Maybe I should just split it in two functions ?

sylviamoss commented on hcl2template/types.packer_config.go:
```diff
@@ -387,23 +384,21 @@ func (cfg *PackerConfig) GetBuilds(opts packer.GetBuildsOptions) ([]packersdk.Bu
-			src, found := cfg.Sources[from.Ref()]
+		for _, srcUsage := range build.Sources {
+			src, found := cfg.Sources[srcUsage.SourceRef]
 			if !found {
```
this check was done before, right?

azr replied:
True ! Better safe than sorry !

sylviamoss commented on hcl2template/types.required_plugins.go:
```diff
@@ -0,0 +1,231 @@
+		}
+
+		nameDiags := checkPluginNameNormalized(name, attr.Expr.Range())
+		diags = append(diags, nameDiags...)
```
wouldn't be better to stop here in case diags has error?

sylviamoss replied:
forget about it! I guess you want to parse the following and append any error

azr replied:
> I guess you want to parse the following and append any error

Yup !

sylviamoss commented on packer/plugin-getter/github/getter.go:
```diff
@@ -0,0 +1,223 @@
+			}
+		}
+		_, _ = buffer.WriteString("]")
+		return ioutil.NopCloser(buffer), nil
```
Will this return only `"[]"` if is not a nominal case?

azr replied:
yes

azr replied:
( leaving it like so, I think this is still okay as it is valid JSON )

sylviamoss commented on packer/plugin-getter/github/getter.go:
```diff
@@ -0,0 +1,223 @@
+		}
+		_, _ = buffer.WriteString("]")
+		return ioutil.NopCloser(buffer), nil
+	}
```
This PR is looking nice! I will continue reviewing the rest tomorrow.

azr commented on command/init.go:
```diff
@@ -0,0 +1,160 @@
+
+func (*InitCommand) Help() string {
+	helpText := `
+Usage: packer init [options] CONFIG
```
yes it will work just like for any other HCL2 command. This is a copy pasta from the RFC; the header for the help for the build command is also a bit confusing:

https://github.com/hashicorp/packer/blob/96f29867ce738facd20aaf0c574c10b83a4e5c65/command/build.go#L378

May be we should do something like:
```suggestion
Usage: packer init [options] [config.pkr.hcl|folder/]
```

azr commented on hcl2template/addrs/plugin.go:
```diff
@@ -0,0 +1,238 @@
+		return "", fmt.Errorf("cannot use multiple consecutive dashes")
+	}
+
+	result, err := idna.Lookup.ToUnicode(given)
```
Kudos to the TF team here I also learned :D

nywilken reviewed (commented):
I just started reviewing the docs to get an idea of how things work first before looking at the code. So far things are off to a good start. We already spoke the installation find. But I have a few small suggestion to start.

azr commented on website/content/docs/commands/init.mdx:
```diff
@@ -0,0 +1,53 @@
+
+The `packer init` Packer command is used to download Packer plugin binaries.
+This is the first command that should be executed when working with a new or
+existing template. This command is always safe to run multiple times. Though
+subsequent runs may give errors, this command will never delete anything.
```
Yes for example. Could be anything. also to mean that if init failed and you already had something working then it should keep on working.

azr replied:
I rephrased that one; is it better like this ? 🙂

azr commented on website/content/docs/configure.mdx:
```diff
@@ -0,0 +1,145 @@
+  default these are 10,000 and 25,000, respectively. Be sure to set a fairly
+  wide range here, since Packer can easily use over 25 ports on a single run.
+
+- `builders`, `commands`, `post-processors`, and `provisioners` are objects
```
Yes

azr replied:
( It's probably not in the docs though 🤔  )

azr replied:
Added the docs !

azr commented on website/content/docs/configure.mdx:
```diff
@@ -0,0 +1,145 @@
+---
+description: |
+  There are various ways to configure Packer. You can not set anything an Packer
+  will use known folders, or you can change things using environment variables.
+page_title: Configuring Packer
```
Yup much better !

azr commented on website/pages/docs/from-1.5/plugin-requirements.mdx:
```diff
@@ -0,0 +1,13 @@
+---
+layout: docs
```
yup had it there but then removed it.

azr commented:
Thanks @nywilken !!

sylviamoss commented on website/content/docs/commands/init.mdx:
```diff
@@ -0,0 +1,58 @@
+-> **Note:** Currently, the plugin getter can only fetch binaries from public
+projects on *Github*.
+
+-> **Note:** Downlading from *Github*, set the `PKR_GITHUB_API_TOKEN` env var in
```
Should we provide a github link with details of how to have a github API token?

azr replied:
Yup adding that now

sylviamoss commented on website/content/docs/plugins/index.mdx:
```diff
@@ -0,0 +1,200 @@
+
+Packer Plugins allow new functionality to be added to Packer without modifying
+the core source code. Packer plugins are able to add new builders, provisioners,
+hooks, and more. In fact, much of Packer itself is implemented by writing
+plugins that are simply distributed with Packer. For example, many of the
+builders, provisioners, and more that ship with Packer are implemented as
+Plugins that are simply hardcoded to load with Packer.
```
I would exclude this part from now on. This won't be true anymore. Maybe just make it a note that currently some plugins are embedded but they will be extracted in the following releases. 🤔 

```suggestion
hooks, and more. 
```

azr replied:
I also wondered about that but I think that the `that ship with Packer` part makes it okay to keep !

sylviamoss commented on website/content/docs/plugins/index.mdx:
```diff
@@ -0,0 +1,200 @@
+plugins](/docs/plugins/creation#developing-plugins) page.
+
+The current official listing of available Packer plugins can be found
+[here](/community-tools#third-party-plugins). This is an incomplete list, and
```
These plugins will work only if they're manually installed. Should we suggest the list here now?

azr replied:
yes !

sylviamoss commented on website/content/docs/plugins/index.mdx:
```diff
@@ -0,0 +1,200 @@
+## Names and Addresses
+Each plugin has two identifiers:
+
+* A `source` address, which is only used when requiring a plugin.
```
Maybe something like
```suggestion
* A `source` address, which is only used when requiring a plugin outside the HashiCorp domain.
```

sylviamoss replied:
If not was suggested, I think this sentence should be different because with or without `source` the plugin is under the required_plugins block so it's requiring a plugin.

sylviamoss commented on website/content/docs/plugins/index.mdx:
```diff
@@ -0,0 +1,200 @@
+  This often is the organization that publishes the plugin. If omitted, this
+  defaults to `hashicorp`, we recommend explicitly setting the namespace.
+
+* **Type:** A short name for the platform or system the plugin manages. The
```
Should an example be added to make it more clear. Like 
githu.com/hashicorp/packer-plugin-amazon source would be github.com/hashicorp/amazon

azr replied:
Yes, there's one just down there.

sylviamoss replied:
Oh yeah! I saw it after commenting on this. Looks great btw

sylviamoss commented on website/content/docs/plugins/index.mdx:
````diff
@@ -0,0 +1,200 @@
+```
+
+The plugin getter will look for plugins located at:
+* github.com/azr/packer-plugin-myawesomecloud
````
oh never mind, here they are!

sylviamoss commented on website/content/docs/plugins/index.mdx:
```diff
@@ -0,0 +1,200 @@
+
+* **Hostname** (optional): The hostname of the location/service that
+  distributes the plugin. If omitted, this defaults to
+  `github.com`, we recommend explicitly setting the hostname.
```
To help people getting not confused if the source is the actual URL for the plugin, maybe we could allow the hostname to be just  `github` too.

azr replied:
Hm that's a good issue 🤔 ; I think that we should keep them as full urls for now.

sylviamoss commented on website/content/partials/plugins/plugin-location.mdx:
```diff
@@ -0,0 +1,54 @@
+Plugins will usually be located in the
+[PACKER_HOME_DIR](/docs/configure#packer-s-home-directory).
+
+* [`packer init`](/commands/init) will install plugins in the **last** directory
```
```suggestion
* [`packer init`](/docs/commands/init) will install plugins in the **last** directory
```

sylviamoss replied:
broken link

azr replied:
Thanks !
